server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchScorePhase.java

```diff
@@ -25,48 +25,44 @@
 import org.apache.lucene.search.Scorer;
 import org.apache.lucene.search.ScorerSupplier;
 import org.apache.lucene.search.Weight;
-import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.fetch.FetchSubPhaseExecutor;
 import org.elasticsearch.search.internal.SearchContext;
 
 import java.io.IOException;
-import java.util.Iterator;
 
 public class FetchScorePhase implements FetchSubPhase {
 
     @Override
-    public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOException {
-        if (context.trackScores() == false || hits.length == 0 ||
-                // scores were already computed since they are needed on the coordinated node to merge top hits
-                context.sort() == null) {
-            return;
+    public FetchSubPhaseExecutor getExecutor(SearchContext context) throws IOException {
+        if (context.trackScores() == false || context.docIdsToLoadSize() == 0 ||
+            // scores were already computed since they are needed on the coordinated node to merge top hits
+            context.sort() == null) {
+            return null;
         }
-
         final IndexSearcher searcher = context.searcher();
         final Weight weight = searcher.createWeight(searcher.rewrite(context.query()), ScoreMode.COMPLETE, 1);
-        Iterator<LeafReaderContext> leafContextIterator = searcher.getIndexReader().leaves().iterator();
-        LeafReaderContext leafContext = null;
-        Scorer scorer = null;
-        for (SearchHit hit : hits) {
-            if (leafContext == null || leafContext.docBase + leafContext.reader().maxDoc() <= hit.docId()) {
-                do {
-                    leafContext = leafContextIterator.next();
-                } while (leafContext == null || leafContext.docBase + leafContext.reader().maxDoc() <= hit.docId());
-                ScorerSupplier scorerSupplier = weight.scorerSupplier(leafContext);
+        return new FetchSubPhaseExecutor() {
+
+            Scorer scorer;
+
+            @Override
+            public void setNextReader(LeafReaderContext readerContext) throws IOException {
+                ScorerSupplier scorerSupplier = weight.scorerSupplier(readerContext);
                 if (scorerSupplier == null) {
-                    throw new IllegalStateException("Can't compute score on document " + hit + " as it doesn't match the query");
+                    scorer = null;
```

Comment:
I think this can safely be turned into an assertion once the `ReaderContext` branch has merged, but should probably stay an Exception for now - I will reinstate it.